File: nodejs-api/src/services/TodoService.js
import { TodoRepository } from '../repositories/TodoRepository.js';
import { ValidationError, NotFoundError } from '../utils/errors.js';
import { createLogger } from '../utils/logger.js';

const logger = createLogger('TodoService');

export class TodoService {
  constructor(todoRepository = null) {
    this.todoRepository = todoRepository || new TodoRepository();
  }

  /**
   * Get all todos with filtering, sorting, and pagination
   */
  async getTodos(options = {}) {
    try {
      logger.info('Getting todos', { options });

      // Validate and sanitize options
      const validatedOptions = this.validateQueryOptions(options);
      
      const result = await this.todoRepository.findAll(validatedOptions);
      
      logger.info('Retrieved todos successfully', {
        count: result.data.length,
        total: result.pagination.total,
        page: result.pagination.page,
      });

      return {
        data: result.data.map(todo => todo.toJSON()),
        ...result.pagination,
      };
    } catch (error) {
      logger.error('Failed to get todos', { error: error.message, options });
      throw error;
    }
  }

  /**
   * Get todo by ID
   */
  async getTodoById(id) {
    try {
      logger.info('Getting todo by ID', { id });

      this.validateId(id);
      
      const todo = await this.todoRepository.findById(id);
      
      if (!todo) {
        logger.warn('Todo not found', { id });
        throw new NotFoundError('Todo');
      }

      logger.info('Retrieved todo successfully', { id, title: todo.title });
      return todo.toJSON();
    } catch (error) {
      logger.error('Failed to get todo by ID', { error: error.message, id });
      throw error;
    }
  }

  /**
   * Create a new todo
   */
  async createTodo(todoData) {
    try {
      logger.info('Creating todo', { title: todoData.title });

      // Validate todo data
      this.validateTodoData(todoData);

      // Sanitize data
      const sanitizedData = this.sanitizeTodoData(todoData);

      const todo = await this.todoRepository.create(sanitizedData);
      
      logger.info('Created todo successfully', { 
        id: todo.id, 
        title: todo.title 
      });

      return todo.toJSON();
    } catch (error) {
      logger.error('Failed to create todo', { 
        error: error.message, 
        todoData 
      });
      throw error;
    }
  }

  /**
   * Update todo by ID
   */
  async updateTodo(id, updates) {
    try {
      logger.info('Updating todo', { id, updates: Object.keys(updates) });

      this.validateId(id);
      this.validateUpdateData(updates);

      // Sanitize updates
      const sanitizedUpdates = this.sanitizeUpdateData(updates);

      if (Object.keys(sanitizedUpdates).length === 0) {
        throw new ValidationError('No valid fields to update');
      }

      const todo = await this.todoRepository.update(id, sanitizedUpdates);
      
      logger.info('Updated todo successfully', { 
        id, 
        title: todo.title 
      });

      return todo.toJSON();
    } catch (error) {
      logger.error('Failed to update todo', { 
        error: error.message, 
        id, 
        updates 
      });
      throw error;
    }
  }

  /**
   * Delete todo by ID
   */
  async deleteTodo(id) {
    try {
      logger.info('Deleting todo', { id });

      this.validateId(id);
      
      await this.todoRepository.delete(id);
      
      logger.info('Deleted todo successfully', { id });
      return true;
    } catch (error) {
      logger.error('Failed to delete todo', { error: error.message, id });
      throw error;
    }
  }

  /**
   * Get todo statistics
   */
  async getTodoStats() {
    try {
      logger.info('Getting todo statistics');

      const stats = await this.todoRepository.getStats();
      
      // Add computed statistics
      const computedStats = {
        ...stats,
        completion_rate: stats.total_todos > 0 
          ? Math.round((stats.completed_todos / stats.total_todos) * 100) 
          : 0,
        last_updated: new Date().toISOString(),
      };

      logger.info('Retrieved todo statistics', computedStats);
      return computedStats;
    } catch (error) {
      logger.error('Failed to get todo statistics', { error: error.message });
      throw error;
    }
  }

  /**
   * Search todos
   */
  async searchTodos(query, options = {}) {
    try {
      logger.info('Searching todos', { query, options });

      if (!query || query.trim().length === 0) {
        throw new ValidationError('Search query is required');
      }

      if (query.length > 255) {
        throw new ValidationError('Search query too long');
      }

      const searchOptions = {
        ...options,
        search: query.trim(),
      };

      return await this.getTodos(searchOptions);
    } catch (error) {
      logger.error('Failed to search todos', { 
        error: error.message, 
        query, 
        options 
      });
      throw error;
    }
  }

  /**
   * Toggle todo completion status
   */
  async toggleTodoCompletion(id) {
    try {
      logger.info('Toggling todo completion', { id });

      const todo = await this.todoRepository.findById(id);
      if (!todo) {
        throw new NotFoundError('Todo');
      }

      const updated = await this.todoRepository.update(id, {
        completed: !todo.completed,
      });

      logger.info('Toggled todo completion', { 
        id, 
        completed: updated.completed 
      });

      return updated.toJSON();
    } catch (error) {
      logger.error('Failed to toggle todo completion', { 
        error: error.message, 
        id 
      });
      throw error;
    }
  }

  /**
   * Bulk operations
   */
  async bulkUpdateTodos(ids, updates) {
    try {
      logger.info('Bulk updating todos', { count: ids.length, updates });

      if (!Array.isArray(ids) || ids.length === 0) {
        throw new ValidationError('IDs array is required');
      }

      // Validate all IDs
      ids.forEach(id => this.validateId(id));
      this.validateUpdateData(updates);

      const sanitizedUpdates = this.sanitizeUpdateData(updates);
      const results = [];

      for (const id of ids) {
        try {
          const todo = await this.todoRepository.update(id, sanitizedUpdates);
          results.push(todo.toJSON());
        } catch (error) {
          if (error instanceof NotFoundError) {
            // Skip missing todos but log the warning
            logger.warn('Todo not found during bulk update', { id });
          } else {
            throw error;
          }
        }
      }

      logger.info('Bulk updated todos', { 
        requested: ids.length, 
        updated: results.length 
      });

      return results;
    } catch (error) {
      logger.error('Failed to bulk update todos', { 
        error: error.message, 
        ids, 
        updates 
      });
      throw error;
    }
  }

  async bulkDeleteTodos(ids) {
    try {
      logger.info('Bulk deleting todos', { count: ids.length });

      if (!Array.isArray(ids) || ids.length === 0) {
        throw new ValidationError('IDs array is required');
      }

      // Validate all IDs
      ids.forEach(id => this.validateId(id));

      const deletedCount = await this.todoRepository.deleteMany(ids);

      logger.info('Bulk deleted todos', { 
        requested: ids.length, 
        deleted: deletedCount 
      });

      return { deleted: deletedCount };
    } catch (error) {
      logger.error('Failed to bulk delete todos', { 
        error: error.message, 
        ids 
      });
      throw error;
    }
  }

  // Private validation methods
  validateId(id) {
    const numId = Number(id);
    if (!Number.isInteger(numId) || numId <= 0) {
      throw new ValidationError('Invalid ID: must be a positive integer');
    }
  }

  validateTodoData(data) {
    if (!data || typeof data !== 'object') {
      throw new ValidationError('Invalid todo data: must be an object');
    }

    if (!data.title || typeof data.title !== 'string' || data.title.trim().length === 0) {
      throw new ValidationError('Title is required');
    }

    if (data.title.length > 255) {
      throw new ValidationError('Title cannot exceed 255 characters');
    }

    if (data.description !== undefined && data.description !== null) {
      if (typeof data.description !== 'string') {
        throw new ValidationError('Description must be a string');
      }
      if (data.description.length > 1000) {
        throw new ValidationError('Description cannot exceed 1000 characters');
      }
    }

    if (data.completed !== undefined && typeof data.completed !== 'boolean') {
      throw new ValidationError('Completed must be a boolean');
    }
  }

  validateUpdateData(data) {
    if (!data || typeof data !== 'object') {
      throw new ValidationError('Invalid update data: must be an object');
    }

    if (Object.keys(data).length === 0) {
      throw new ValidationError('At least one field must be provided');
    }

    const allowedFields = ['title', 'description', 'completed'];
    const invalidFields = Object.keys(data).filter(key => !allowedFields.includes(key));
    
    if (invalidFields.length > 0) {
      throw new ValidationError(`Invalid fields: ${invalidFields.join(', ')}`);
    }

    // Validate individual fields if present
    if ('title' in data) {
      if (!data.title || typeof data.title !== 'string' || data.title.trim().length === 0) {
        throw new ValidationError('Title cannot be empty');
      }
      if (data.title.length > 255) {
        throw new ValidationError('Title cannot exceed 255 characters');
      }
    }

    if ('description' in data && data.description !== null) {
      if (typeof data.description !== 'string') {
        throw new ValidationError('Description must be a string');
      }
      if (data.description.length > 1000) {
        throw new ValidationError('Description cannot exceed 1000 characters');
      }
    }

    if ('completed' in data && typeof data.completed !== 'boolean') {
      throw new ValidationError('Completed must be a boolean');
    }
  }

  validateQueryOptions(options) {
    const defaults = {
      page: 1,
      per_page: 20,
      sort: 'created_at',
      order: 'desc',
    };

    const validated = { ...defaults, ...options };

    // Validate page
    if (validated.page < 1) {
      validated.page = 1;
    }

    // Validate per_page
    if (validated.per_page < 1 || validated.per_page > 100) {
      validated.per_page = 20;
    }

    // Validate sort field
    const allowedSortFields = ['id', 'title', 'completed', 'created_at', 'updated_at'];
    if (!allowedSortFields.includes(validated.sort)) {
      validated.sort = 'created_at';
    }

    // Validate order
    if (!['asc', 'desc'].includes(validated.order)) {
      validated.order = 'desc';
    }

    // Validate search
    if (validated.search && validated.search.length > 255) {
      throw new ValidationError('Search query too long');
    }

    return validated;
  }

  sanitizeTodoData(data) {
    return {
      title: data.title.trim(),
      description: data.description?.trim() || null,
      completed: Boolean(data.completed),
    };
  }

  sanitizeUpdateData(data) {
    const sanitized = {};

    if ('title' in data) {
      sanitized.title = data.title.trim();
    }

    if ('description' in data) {
      sanitized.description = data.description?.trim() || null;
    }

    if ('completed' in data) {
      sanitized.completed = Boolean(data.completed);
    }

    return sanitized;
  }
}

export default TodoService;

File: golang-api/internal/services/todo_service.go
package services

import (
	"fmt"
	"log/slog"
	"strings"
	"time"

	"github.com/centroidsol/todo-api/internal/models"
	"github.com/centroidsol/todo-api/internal/repository"
)

type TodoService interface {
	GetTodos(params models.QueryParams) (*models.PaginatedResponse, error)
	GetTodoByID(id int) (*models.Todo, error)
	CreateTodo(req models.CreateTodoRequest) (*models.Todo, error)
	UpdateTodo(id int, req models.UpdateTodoRequest) (*models.Todo, error)
	DeleteTodo(id int) error
	DeleteAllTodos() error
	GetTodoStats() (map[string]interface{}, error)
}

type todoService struct {
	repo   repository.TodoRepository
	logger *slog.Logger
}

func NewTodoService(repo repository.TodoRepository, logger *slog.Logger) TodoService {
	return &todoService{
		repo:   repo,
		logger: logger,
	}
}

func (s *todoService) GetTodos(params models.QueryParams) (*models.PaginatedResponse, error) {
	s.logger.Info("Getting todos", "params", params)

	// Validate and set defaults
	if params.Page < 1 {
		params.Page = 1
	}
	if params.PerPage < 1 || params.PerPage > 100 {
		params.PerPage = 20
	}
	if params.Sort == "" {
		params.Sort = "created_at"
	}
	if params.Order == "" {
		params.Order = "desc"
	}

	// Validate sort field
	validSortFields := []string{"id", "title", "completed", "created_at", "updated_at"}
	if !contains(validSortFields, params.Sort) {
		return nil, fmt.Errorf("invalid sort field: %s", params.Sort)
	}

	// Validate order
	if params.Order != "asc" && params.Order != "desc" {
		return nil, fmt.Errorf("invalid order: %s", params.Order)
	}

	todos, total, err := s.repo.GetAll(params)
	if err != nil {
		s.logger.Error("Failed to get todos", "error", err)
		return nil, fmt.Errorf("failed to get todos: %w", err)
	}

	totalPages := (total + params.PerPage - 1) / params.PerPage

	response := &models.PaginatedResponse{
		Data:       todos,
		Total:      total,
		Page:       params.Page,
		PerPage:    params.PerPage,
		TotalPages: totalPages,
	}

	s.logger.Info("Retrieved todos successfully", "count", len(todos), "total", total)
	return response, nil
}

func (s *todoService) GetTodoByID(id int) (*models.Todo, error) {
	s.logger.Info("Getting todo by ID", "id", id)

	if id <= 0 {
		return nil, fmt.Errorf("invalid todo ID: %d", id)
	}

	todo, err := s.repo.GetByID(id)
	if err != nil {
		s.logger.Error("Failed to get todo by ID", "id", id, "error", err)
		return nil, fmt.Errorf("failed to get todo: %w", err)
	}

	if todo == nil {
		s.logger.Warn("Todo not found", "id", id)
		return nil, nil
	}

	s.logger.Info("Retrieved todo successfully", "id", id, "title", todo.Title)
	return todo, nil
}

func (s *todoService) CreateTodo(req models.CreateTodoRequest) (*models.Todo, error) {
	s.logger.Info("Creating todo", "title", req.Title)

	// Validate request
	if err := s.validateCreateRequest(req); err != nil {
		return nil, err
	}

	// Create todo model
	todo := &models.Todo{
		Title:       strings.TrimSpace(req.Title),
		Description: req.Description,
		Completed:   req.Completed,
		CreatedAt:   time.Now(),
		UpdatedAt:   time.Now(),
	}

	// Trim description if provided
	if todo.Description != nil {
		trimmed := strings.TrimSpace(*todo.Description)
		if trimmed == "" {
			todo.Description = nil
		} else {
			todo.Description = &trimmed
		}
	}

	if err := s.repo.Create(todo); err != nil {
		s.logger.Error("Failed to create todo", "error", err)
		return nil, fmt.Errorf("failed to create todo: %w", err)
	}

	s.logger.Info("Created todo successfully", "id", todo.ID, "title", todo.Title)
	return todo, nil
}

func (s *todoService) UpdateTodo(id int, req models.UpdateTodoRequest) (*models.Todo, error) {
	s.logger.Info("Updating todo", "id", id)

	if id <= 0 {
		return nil, fmt.Errorf("invalid todo ID: %d", id)
	}

	// Validate request
	if err := s.validateUpdateRequest(req); err != nil {
		return nil, err
	}

	// Check if todo exists
	exists, err := s.repo.Exists(id)
	if err != nil {
		s.logger.Error("Failed to check todo existence", "id", id, "error", err)
		return nil, fmt.Errorf("failed to check todo existence: %w", err)
	}

	if !exists {
		s.logger.Warn("Todo not found for update", "id", id)
		return nil, nil
	}

	// Build updates map
	updates := make(map[string]interface{})

	if req.Title != nil {
		trimmed := strings.TrimSpace(*req.Title)
		updates["title"] = trimmed
	}

	if req.Description != nil {
		trimmed := strings.TrimSpace(*req.Description)
		if trimmed == "" {
			updates["description"] = nil
		} else {
			updates["description"] = trimmed
		}
	}

	if req.Completed != nil {
		updates["completed"] = *req.Completed
	}

	// Perform update
	todo, err := s.repo.Update(id, updates)
	if err != nil {
		s.logger.Error("Failed to update todo", "id", id, "error", err)
		return nil, fmt.Errorf("failed to update todo: %w", err)
	}

	s.logger.Info("Updated todo successfully", "id", id)
	return todo, nil
}

func (s *todoService) DeleteTodo(id int) error {
	s.logger.Info("Deleting todo", "id", id)

	if id <= 0 {
		return fmt.Errorf("invalid todo ID: %d", id)
	}

	// Check if todo exists
	exists, err := s.repo.Exists(id)
	if err != nil {
		s.logger.Error("Failed to check todo existence", "id", id, "error", err)
		return fmt.Errorf("failed to check todo existence: %w", err)
	}

	if !exists {
		s.logger.Warn("Todo not found for deletion", "id", id)
		return fmt.Errorf("todo with id %d not found", id)
	}

	if err := s.repo.Delete(id); err != nil {
		s.logger.Error("Failed to delete todo", "id", id, "error", err)
		return fmt.Errorf("failed to delete todo: %w", err)
	}

	s.logger.Info("Deleted todo successfully", "id", id)
	return nil
}

func (s *todoService) DeleteAllTodos() error {
	s.logger.Info("Deleting all todos")

	if err := s.repo.DeleteAll(); err != nil {
		s.logger.Error("Failed to delete all todos", "error", err)
		return fmt.Errorf("failed to delete all todos: %w", err)
	}

	s.logger.Info("Deleted all todos successfully")
	return nil
}

func (s *todoService) GetTodoStats() (map[string]interface{}, error) {
	s.logger.Info("Getting todo statistics")

	// Get all todos to calculate stats
	params := models.QueryParams{
		Page:    1,
		PerPage: 1000, // Get a large number to get all todos
		Sort:    "created_at",
		Order:   "desc",
	}

	response, err := s.GetTodos(params)
	if err != nil {
		return nil, err
	}

	todos := response.Data.([]models.Todo)

	stats := map[string]interface{}{
		"total_todos":     response.Total,
		"completed_todos": 0,
		"pending_todos":   0,
	}

	for _, todo := range todos {
		if todo.Completed {
			stats["completed_todos"] = stats["completed_todos"].(int) + 1
		} else {
			stats["pending_todos"] = stats["pending_todos"].(int) + 1
		}
	}

	s.logger.Info("Retrieved todo statistics", "stats", stats)
	return stats, nil
}

func (s *todoService) validateCreateRequest(req models.CreateTodoRequest) error {
	if strings.TrimSpace(req.Title) == "" {
		return fmt.Errorf("title is required")
	}

	if len(req.Title) > 255 {
		return fmt.Errorf("title cannot exceed 255 characters")
	}

	if req.Description != nil && len(*req.Description) > 1000 {
		return fmt.Errorf("description cannot exceed 1000 characters")
	}

	return nil
}

func (s *todoService) validateUpdateRequest(req models.UpdateTodoRequest) error {
	if req.Title != nil {
		if strings.TrimSpace(*req.Title) == "" {
			return fmt.Errorf("title cannot be empty")
		}
		if len(*req.Title) > 255 {
			return fmt.Errorf("title cannot exceed 255 characters")
		}
	}

	if req.Description != nil && len(*req.Description) > 1000 {
		return fmt.Errorf("description cannot exceed 1000 characters")
	}

	return nil
}

func contains(slice []string, item string) bool {
	for _, s := range slice {
		if s == item {
			return true
		}
	}
	return false
}
